internal/compiler/passes/resolving/remove_noop.rs

```diff
@@ -0,0 +1,110 @@
+/// will therefore not modify the last Expression in the Vec, as that forms the result value of the
+/// CodeBlock itself.
+pub fn remove_from_codeblock(
+    code_block: &mut Vec<(SyntaxNode, Expression)>,
+    diagnostics: &mut BuildDiagnostics,
+) {
+    if code_block.len() > 1 {
+        // In a code block, only the last expression returns a value.
+        // Therefore all other expressions inside the block are only useful if they have side
+        // effects.
+        //
+        // Remove all expressions without side effects (except for the last one) and emit a
+        // warning.
+        //
+        // Note: Iterate over the indices in reverse, so that all to-be-iterated indices remain
+        // valid when removing items from the vector.
+        for index in (0..(code_block.len() - 1)).rev() {
+            let (node, expression) = &code_block[index];
+            if without_side_effects(expression) {
+                diagnostics.push_warning("Expresssion has no effect!".to_owned(), node);
+                code_block.remove(index);
+            }
+        }
+    }
+}
+
+/// Returns whether the expression is certain to be without side effects.
+/// This function is conservative and may still return `false`, even if a given expression
+/// is without side effects.
+/// It is only guaranteed that if this function returns `true`, the expression definitely does not
+/// contain side effects.
+fn without_side_effects(expression: &Expression) -> bool {
+    match expression {
+        Expression::Condition { condition, true_expr, false_expr } => {
+            without_side_effects(condition)
+                && without_side_effects(true_expr)
+                && without_side_effects(false_expr)
+        }
+        Expression::NumberLiteral(_, _) => true,
+        Expression::StringLiteral(_) => true,
+        Expression::BoolLiteral(_) => true,
+        Expression::CodeBlock(expressions) => expressions.iter().all(without_side_effects),
+        Expression::FunctionParameterReference { .. } => true,
+        // Invalid and uncompiled expressions are unknown at this point, so default to
+        // `false`, because they may have side-efffects.
+        Expression::Invalid => false,
+        Expression::Uncompiled(_) => false,
+        // A property reference may cause re-evaluation of a property, which may result in
+        // side effects
+        Expression::PropertyReference(_) => false,
+        Expression::ElementReference(_) => false,
+        Expression::RepeaterIndexReference { .. } => true,
+        Expression::RepeaterModelReference { .. } => true,
+        Expression::StoreLocalVariable { .. } => false,
+        Expression::ReadLocalVariable { .. } => true,
+        Expression::StructFieldAccess { base, name: _ } => without_side_effects(&*base),
+        Expression::ArrayIndex { array, index } => {
+            without_side_effects(&*array) && without_side_effects(&*index)
```

Comment:
if i do something like

```slint
clicked => {
   some_model[something_with_side_effect()];
}
```

The `some_model[]` thing is still useless and we could still warn about it, don't you think?

Same with `something_with_side_effect().foobar;` and so on.